modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/SessionImpl.java

```diff
@@ -364,6 +318,12 @@ public static <T> T await(CompletionStage<T> stage) {
         }
     }
 
+    private void closeInternal() {
+        if (closed.compareAndSet(false, true)) {
+            busyLock.block();
```

Comment:
Second thread will not wait for `busyLock.block()` and proceeds with `qryProc.closeSession(sessionId)`.

Reply:
It's ok. A busyLock is used to prevent the session being used after cancellation. Subsequent call to qryProc.cancel() will not have any effect according to the javadoc